p4-16/spec/P4-16-draft-spec.mdk

```diff
@@ -2647,196 +2645,191 @@ X.v1  // reference to v1
 v1    // error - v1 is not in the top-level namespace
 ~ End P4Example
 
-An ```enum``` value can only be compared for equality/difference using ```== and !=```. ```enum```
-values cannot be cast to or from any other types.
+Similar to errors, ```enum``` expressions only support equality (`==`)
+and inequality (`!=`) comparisons. Expressions whose type is an `enum`
+cannot be cast to or from any other type.
 
-When ```enum``` values appear in the control-plane API the compiler
-back-end has to choose a suitable serialization data type and
+Note that if an ```enum``` value appears in the control-plane API, the
+compiler must select a suitable serialization data type and
 representation.
 
-## Expressions on boolean values { #sec-bool-exprs }
+## Expressions on boolean expressions { #sec-bool-exprs }
 
-The following operations are provided on boolean values:
-- And, designated by ```&&```
-- Or designated by ```||```
-- Negation, designated by ```!```
-- Equality tests (```==``` and ```!=```)
+The following operations are provided on boolean expressions:
+- And, denoted by ```&&```,
+- Or denoted by ```||```,
+- Negation, denoted by ```!```, and
+- Equality and inequality tests, denoted by ```==``` and ```!=``` respectively.
 
-Operator precedence is similar to C. Operator evaluation is
-short-circuit.
+The precedence of these operators is similar to C and uses short-circuit evaluation.
 
-There are no implicit casts from bit-strings to booleans or
-vice-versa. As a consequence, a C program fragment such as:
+P4 does not implicit cast from bit-strings to booleans and
+vice versa. As a consequence, a program is valid in a language like C such as,
 
 ~ Begin P4Example
 if (x) ...
 ~ End P4Example
 
-(for x an integer base type) must be written in P4 as:
+(where x has an integer type) must instead be written in P4 as:
 
 ~ Begin P4Example
 if (x != 0) ...
 ~ End P4Example
 
-(see also the discussion on infinite-precision types and implicit
-casts in Section [#sec-implicit-casts] for how the 0 in this
-expression is evaluated).
-
-### The conditional operator
+See the discussion on infinite-precision types and implicit casts
+in Section [#sec-implicit-casts] for details on how the `0` in this
+expression is evaluated.
 
-The ```?:``` expression behaves as in C, e.g.:
+### Conditional operator
 
-~ Begin P4Example
-(x == 0) ? e0 : e1;
-~ End P4Example
+A conditional expression of the form ```e1 ? e2 : e3``` behaves the
+same as in languages like C. The expression `e1` is evaluated first,
+and either `e2` or `e3` is evaluated depending on the result.
```

Comment:
Added "As described above, ..."

In general, this section is repetitive but I think that's not terrible. We're sacrificing readability for the end-to-end reader, but making it easier to flip open to a random subsection and get a full picture of how particular features behave.